### DISCUSSION TITLE
Converting datetime to UTC and ISO format ends up with Z instead of +00:00

### DISCUSSION THREAD

**Discussion by houmie** [fastapi/fastapi#9887] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
expiry_date = (device.expires_at.replace(tzinfo=pytz.UTC)).isoformat()
return {
        "expires_at": expiry_date
    }
```

### Description

+00:00" is a valid ISO 8601 timezone designation for UTC. But this seems to have changed to Z. Is there a way to change this?

In fastapi==0.95.0 we created our ISO date like this and returned it as json:

```
expiry_date = (device.expires_at.replace(tzinfo=pytz.UTC)).isoformat()
return {
        "expires_at": expiry_date
    }
```

The unit test was expecting a dateformat like this:
`2023-07-16T06:26:30.769459+00:00`

```
assert response.json()["expires_at"] == date.strftime(
    "%Y-%m-%dT%H:%M:%S.%f+00:00"
)
```

But now with fastapi==0.100.0 the format has changed:
`2023-07-16T06:26:30.769459Z`

Is there a way to change this back to +00:00?

Many Thanks

### Operating System

Linux

### Operating System Details

MacOS and Linux

### FastAPI Version

0.100.0

### Python Version

3.11.4

**Comment by jgould22** [comment]
The major change from FastAPI < 100 to FastAPI 100.0 is that Pydantic was upgraded to 2.0.

Recreating something like your sample below 
```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from datetime import datetime, date
import uvicorn
import pytz
import pytest

app = FastAPI()

@app.get("/")
async def root():
    test_date = datetime.now()
    test_date_tz = (datetime.now().replace(tzinfo=pytz.UTC)).isoformat()
    return {
        "date": test_date,
        "date_tz": test_date_tz
    }

client = TestClient(app)

def test_date():
    response = client.get("/")
    print( response.json())
    test_date = datetime.strptime(response.json()["date_tz"], "%Y-%m-%dT%H:%M:%S.%f+00:00").date()
    print(test_date)
``` 

```javascript
{'date': '2023-07-16T12:04:18.439371', 'date_tz': '2023-07-16T12:04:18.439375+00:00'}
```
I do not get a Z in my response. 

Perhaps it is related to the types that are not included in your sample code but we probably need a more complete piece of code to see what might be happening.

**Reply by houmie** [reply]
@jgould22 

Jordan,

Apologies for my late reply. We tested this further and the problem is still there.

You have forgotten to use a Pydantic model for the response_model. That's on me, though, as I didn't explain it well enough in my previous post.

I have adjusted your example, and you will see that your test is now failing. The date has a "Z" at the end, representing the zero offset from UTC. This has significant implications for us, as all our iOS devices will fail to connect to the API if we upgrade FastAPI.

What do you think? It seems like a bug to me.

```
from datetime import datetime
from typing import Optional

import pytz
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel

app = FastAPI()


class VerifyUserOut(BaseModel):
    expires_at: Optional[datetime] = None


@app.get(
    "/",
    status_code=200,
    response_model=VerifyUserOut,
    responses={200: {"model": VerifyUserOut}},
)
async def root():
    expires_at = (datetime.utcnow().replace(tzinfo=pytz.UTC)).isoformat()
    return {"expires_at": expires_at}


client = TestClient(app)


def test_date():
    response = client.get("/")
    print(response.json())
    test_date = datetime.strptime(
        response.json()["expires_at"], "%Y-%m-%dT%H:%M:%S.%f+00:00"
    ).date()
    print(test_date)
```

**Reply by jgould22** [reply]
Add this to your `VerifyUserOut` model

```python 

    @field_serializer('expires_at')
    def serialize_expires_at(self, expires_at: datetime, _info):
        return expires_at.strftime("%Y-%m-%dT%H:%M:%S.%f+00:00")

```

more info here https://docs.pydantic.dev/2.0/usage/serialization/#custom-serializers

**Reply by houmie** [reply]
Ah thank you.  This looks more like a workaround to me though. What do you think?  We didn't have to do this before, now I need to add this to the entire pydantic models, because datetime is suddenly formatted with a z at the end. It doesn't feel clean. I think it's fair to raise this as a bug. At least it can be looked into and rejected in worse case scenario. Thanks

**Reply by houmie** [reply]
@tiangolo  What do you think? Do you agree this is a bug? It doesn't seem like a clean solution to me, compared to how FastAPI < 100 was working before in this regard. Thanks

**Reply by JavierSanchezCastro** [reply]
After reviewing the code, and as @jgould22 said, it's pydantic side.

Also, I see that there is an issue closed a few minutes ago: https://github.com/pydantic/pydantic/issues/6735

Some other issues that report problems with datetime: https://github.com/pydantic/pydantic/issues/7039 https://github.com/pydantic/pydantic/issues/6592 https://github.com/pydantic/pydantic/issues/6576

**Reply by houmie** [reply]
Thanks for the info. 👍🏼 You are right it's a pydantic issue. It seems indeed quite a bit of mess going by the number of reports about this. At least the workaround works on Fastapi, which is great.  I'm keen to see how this develops and if they will be patching this or not.
